File: shared/protocol/payloads/Payloads.cs
namespace WatchParty.Shared.Protocol.Payloads;





public record IptvContentDescriptor(
    string ContentType,          
    string StreamId,
    string? ContainerExtension,
    string Title)
{
    public string ContentKey => $"{ContentType}|{StreamId}|{ContainerExtension ?? string.Empty}";
}



public record JoinRoomPayload(string RoomCode, string Role);

public record SetContentPayload(IptvContentDescriptor Descriptor);

public record PlayPayload(long PositionMs, long ClientTimestampMs);

public record PausePayload(long PositionMs);

public record SeekPayload(long TargetPositionMs);

public record BufferingStallPayload(long PositionMs);

public record PingPayload(long ClientTimestampMs);
public record SetPlaybackSpeedPayload(double Speed);



public record RoomJoinedPayload(
    string RoomCode,
    string Role,
    bool GuestPresent,
    IptvContentDescriptor? ContentDescriptor,
    long ServerTimestampMs,
    double PlaybackRate = 1.0);

public record RoomGuestJoinedPayload(long ServerTimestampMs);

public record RoomGuestLeftPayload(long ServerTimestampMs, int GracePeriodSeconds);

public record RoomGuestReconnectedPayload(long ServerTimestampMs);

public record RoomClosedPayload(string Reason, long ServerTimestampMs);

public record RoomContentSetPayload(IptvContentDescriptor Descriptor, long ServerTimestampMs);

public record ErrorPayload(string Code, string Message, long ServerTimestampMs);


public record PlaybackPlayPayload(long PositionMs, long ServerTimestampMs, int HostRttMs, int SeqNo, double PlaybackRate = 1.0);


public record PlaybackPausePayload(long PositionMs, long ServerTimestampMs, int SeqNo);



public record PlaybackSeekPayload(long TargetPositionMs, long ServerTimestampMs, int SeqNo, bool IsPlaying);


public record PlaybackStateSyncPayload(long HostPositionMs, bool IsPlaying, long ServerTimestampMs, int SeqNo, double PlaybackRate = 1.0);

public record PlaybackSpeedPayload(double Speed, long ServerTimestampMs);

public record BufferingStallBroadcastPayload(string Role, long PositionMs, long ServerTimestampMs, int EpisodeId);

public record BufferingResumePayload(long ServerTimestampMs, long ResumePositionMs, int EpisodeId, bool IsPlaying);

public record PlayerReadyPayload(bool BothReady, string ReadyRole, long ServerTimestampMs, string ContentKey);

public record PongPayload(long ClientTimestampMs, long ServerTimestampMs);

File: client_samsung_tv/src/protocol/payloads.ts
export type IptvDescriptorType = 'live' | 'movie' | 'episode';
export type RoomRole = 'host' | 'guest';

export interface IptvContentDescriptor {
  contentType: IptvDescriptorType;
  streamId: string;
  containerExtension?: string | null;
  title: string;
  contentKey?: string;
}

export interface RoomJoinedPayload {
  roomCode: string;
  role: RoomRole;
  guestPresent: boolean;
  contentDescriptor?: IptvContentDescriptor | null;
  serverTimestampMs: number;
  playbackRate?: number;
}

export interface RoomGuestJoinedPayload {
  serverTimestampMs: number;
}

export interface RoomGuestLeftPayload {
  serverTimestampMs: number;
  gracePeriodSeconds: number;
}

export interface RoomGuestReconnectedPayload {
  serverTimestampMs: number;
}

export interface RoomHostAwayPayload {
  serverTimestampMs: number;
  gracePeriodSeconds: number;
}

export interface RoomHostReconnectedPayload {
  serverTimestampMs: number;
}

export interface RoomClosedPayload {
  reason: string;
  serverTimestampMs: number;
}

export interface RoomContentSetPayload {
  descriptor: IptvContentDescriptor;
  serverTimestampMs: number;
}

export interface ErrorPayload {
  code: string;
  message: string;
  serverTimestampMs: number;
}

export interface PlaybackPlayPayload {
  positionMs: number;
  serverTimestampMs: number;
  hostRttMs: number;
  seqNo?: number;
  playbackRate?: number;
}

export interface PlaybackPausePayload {
  positionMs: number;
  serverTimestampMs: number;
  seqNo?: number;
}

export interface PlaybackSeekPayload {
  targetPositionMs: number;
  serverTimestampMs: number;
  seqNo?: number;
  isPlaying?: boolean;
}

export interface PlaybackStateSyncPayload {
  hostPositionMs: number;
  isPlaying: boolean;
  serverTimestampMs: number;
  seqNo?: number;
  playbackRate?: number;
}

export interface PlaybackSpeedPayload {
  speed: number;
  serverTimestampMs: number;
}

export interface PongPayload {
  clientTimestampMs: number;
  serverTimestampMs: number;
}

export interface BufferingStallBroadcastPayload {
  episodeId?: number;
  role: string;
  positionMs: number;
  serverTimestampMs: number;
}

export interface PlayerReadyPayload {
  bothReady: boolean;
  readyRole: RoomRole;
  serverTimestampMs: number;
  contentKey: string;
}

export interface BufferingResumePayload {
  episodeId?: number;
  serverTimestampMs: number;
  resumePositionMs: number;
  isPlaying?: boolean;
}

export function contentKeyOf(descriptor: IptvContentDescriptor): string {
  return `${descriptor.contentType}|${descriptor.streamId}|${descriptor.containerExtension ?? ''}`;
}

export function normalizeDescriptor(descriptor: IptvContentDescriptor): IptvContentDescriptor {
  return {
    contentType: descriptor.contentType,
    streamId: String(descriptor.streamId),
    containerExtension: descriptor.containerExtension ?? null,
    title: descriptor.title,
  };
}
